File: Hash Table/249M. Group Shifted Strings.py
'''
249M. Group Shifted Strings
Hash Table

Description:
Given a string, we can "shift" each of its letter to its successive letter, for example: "abc" -> "bcd". 
We can keep "shifting" which forms the sequence: "abc" -> "bcd" -> ... -> "xyz" or "adg -> beh -> cfi ..."
Given a list of strings which contains only lowercase alphabets, group all strings that belong to the same shifting sequence.

Example:
given: ["abc", "bcd", "acef", "xyz", "az", "ba", "a", "z"],
A solution is:
[
  ["abc","bcd","xyz"],
  ["az","ba"],
  ["acef"],
  ["a","z"]
]
'''

# Solution:
class Solution(object):
    def groupStrings(self, strs):
        # Use function ord() to convert alphabet to integer. ord('a') -> 97, chr(97) -> 'a'
        # We use ord('a') - 97 to convert 'a' to 0.
        # Note that 'abc' -> 0,1,2, 'bcd' -> 1,2,3 and 'xyz' -> 23,24,25. If we subtract the first number from the list,
        #   we can get  'abc' -> 0,1,2, 'bcd' -> 0,1,2 and 'xyz' -> 0,1,2
        # Notice that the successive letter of 'z' is 'a'. 'az' -> 0,25 and 'ba' -> 1,0.
        #   If we subtract 1 from 1,0, we get 0, -1. In this case, we can take the modulus of 26.
        #   0 % 26 = 0, -1 % 26 = 25.

        # The basic idea to convert the string to a list of integers.
        # Check if the sequence of the integers has been seen. If not, create a new key. If so, save it to existing key.

        result = {}
        for word in strs:
            word_converted = self.convertWord(word)
            # print(word_converted)
            if word_converted not in result:
                result[word_converted] = [word]
            else:
                result[word_converted] = result[word_converted] + [word]
        return result.values()

    def convertWord(self, word):
        L_word = []
        for char in word:
            L_word.append(ord(char))
        L_word = [(num - L_word[0]) % 26 for num in L_word]
        L_str = ''
        for num in L_word:
            L_str += str(num)
        return L_str
    

File: Hash Table/249M. Group Shifted Strings.cpp
/*
Description:
Given a string, we can "shift" each of its letter to its successive letter, for example: "abc" -> "bcd".
We can keep "shifting" which forms the sequence: "abc" -> "bcd" -> ... -> "xyz" or "adg -> beh -> cfi ..."
Given a list of strings which contains only lowercase alphabets, group all strings that belong to the same shifting sequence.
Example:
given: ["abc", "bcd", "acef", "xyz", "az", "ba", "a", "z"],
A solution is:
[
  ["abc","bcd","xyz"],
  ["az","ba"],
  ["acef"],
  ["a","z"]
]
*/
//
// Created by Shiyi Yang on 2021-03-11.
//
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupStrings(vector<string>& strings) {
        vector<vector<string>> res;
        unordered_map<string, vector<string>> m;
        for (auto str:strings) {
            m[convertString(str)].push_back(str);
        }
        unordered_map<string, vector<string>>::iterator it=m.begin();
        while (it!=m.end()) {
            cout << (*it).first << endl;
            res.push_back(it->second);
            it++;
        }
        return res;
    }

    string convertString(string str) {
        vector<int> v;
        string res;
        int num1, num2;
        /*
        for (int i=0; i<str.size(); ++i) {
            num1 = (int)str[i] - 97;
            v.push_back(num1);
        }
        for (int i=0; i<v.size(); ++i) {
            num2 = v[i] - v[0];
            if (num2 < 0) {
                num2 +=  26;
            }
            res.insert(i, to_string(num2));
        }
        */
        
        int num0 = (int)str[0] - 97;
        for (int i=0; i<str.size(); ++i) {
            num1 = (int)str[i] - 97 - num0;
            if (num1 < 0) {
                num1 += 26;
            }
            res.insert(i, to_string(num1));
        }
      
        return res;
    }
};

void print_vector(vector<vector<string>> v){
    cout << "[";
    for(int i = 0; i<v.size(); i++){
        cout << "[";
        for(int j = 0; j <v[i].size(); j++){
            cout << v[i][j] << ", ";
        }
        cout << "], ";
    }
    cout << "]"<<endl;
}

int main() {
    Solution* obj = new Solution();
    vector<string> vec = {"abc", "bcd", "acef", "xyz", "az", "ba", "a", "z"};
    print_vector(obj->groupStrings(vec));  //[[a, z, ], [acef, ], [az, ba, ], [abc, bcd, xyz, ], ]

    return 0;
}
